File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
read = sys.stdin.readline

n, m = map(int, read().split())
board = [list(read().rstrip()) for _ in range(n)]
di = [-1, 0, 1, 0]
dj = [0, 1, 0, -1]


def bfs():
    queue = [(0, 0)]
    move = 1
    while queue:
        temp = []
        for i, j in queue:
            for d in range(4):
                ni = i + di[d]
                nj = j + dj[d]
                if i == n - 1 and j == m - 1:  # arrive
                    return move
                if 0 <= ni < n and 0 <= nj < m:
                    if board[ni][nj] == '1':
                        temp.append((ni, nj))
                        board[ni][nj] = '0'
        queue = temp
        move += 1


print(bfs())

File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const input = require('fs').readFileSync(process.platform === "linux" ? "/dev/stdin" : "./sample.txt").toString().trim().split("\n");

const [n, m] = input[0].split(" ").map(Number);
const board = []; // 방문 후 2
const delta = [[-1, 0], [0, 1], [1, 0], [0, -1]];
let queue = [];
let cnt, nx, ny;

for (let i = 1; i < 1 + n; i++) {
  board.push(Array.from(input[i]).map(Number));
}

queue.push([0, 0]);
board[0][0] = 2;

(() => {
  cnt = 1;
  while (queue.length > 0) {
    let len = queue.length;
    for (let i = 0; i < len; i++) {
      const [x, y] = queue.shift();
      for (let [dx, dy] of delta) {
        nx = x + dx;
        ny = y + dy;
        if (0 <= nx && nx < n && 0 <= ny && ny < m && board[nx][ny] === 1) {
          if (nx === n - 1 && ny === m - 1) {
            console.log(cnt + 1);
            return;
          }
          queue.push([nx, ny]);
          board[nx][ny] = 2;
        }
      }
    }

    cnt += 1;
  }

})();